Extend Fedora to support kernel updates on chromebooks

dorindabassey opened the pull request:
Added a kernel-install script that creates an intramfs and FIT Image (including the ramdisk)
then the script flashes the kernel to the right partition in order to support kernel updates on the chromebook

Signed-off-by: Dorinda Bassey <dbassey@redhat.com>

martinezjavier commented on 96-chromebook.install:
```diff
@@ -0,0 +1,62 @@
+
+case "$COMMAND" in
+    add)
+        echo "Install required packages"
```
I don't think other kernel-install scripts print an output ?

martinezjavier commented on 96-chromebook.install:
```diff
@@ -0,0 +1,62 @@
+case "$COMMAND" in
+    add)
+        echo "Install required packages"
+        dnf install lz4 vboot-utils uboot-tools -y
```
I would prefer if this script instead was shipped let's say as a part of vboot-utils? or maybe a vboot-utils-kernel-install sub-package and pulled these as dependencies ?

martinezjavier replied:
just to make it clear, I don't think that the script should install packages. In any case if you want to ship this with `chromebook-setup.sh`, then these packages should be added as dependencies in the script itself using `ensure_command`.

dorindabassey replied:
like do you mean submitting a patch to include the 96-chromebook.install script in vboot-utils ustream?

martinezjavier replied:
yes. But I would ship it as a separate sub-package since it may be that someone wants to have the vboot-utils tools in their host machine for cross-compile and then ship to a target device. So a `vboot-kernel-install` or something sub-package would make sense.

That way you won't need to manually install the dependencies for this since it would be handled by the package manager.

dorindabassey replied:
oh I get what you mean now, thanks.

martinezjavier commented on 96-chromebook.install:
```diff
@@ -0,0 +1,62 @@
+        i2c-qcom-geni i2c-qup icc-osm-l3 qcom-spmi-pmic phy-qcom-qmp phy-qcom-qusb2 \
+        phy-qcom-usb-hs qcom_aoss qcom-apcs-ipc-mailbox llcc-qcom nvmem_qfprom smem \
+        smp2p dwc3-qcom \
+        " "$BOOT_DIR_ABS/$INITRD" --kver "$KERNEL_VERSION" --kmoddir /lib/modules/"${KERNEL_VERSION}"
```
I don't think all this has to be done here? this be handled by the dracut kernel-install plugin (`/usr/lib/kernel/install.d/50-dracut.install`) already.

dorindabassey replied:
this was the list of required modules provided by enric for the initramfs, so this section will regenerate an initramfs to suit the chromebook.

martinezjavier replied:
yes, that makes sense for creating an initial initrd but that's not needed when updating a kernel package. Since dracut would run on the target machine in that case and would figure out what modules need to be included.

dorindabassey replied:
Okay, I see. Thanks

martinezjavier commented on 96-chromebook.install:
```diff
@@ -0,0 +1,62 @@
+        " "$BOOT_DIR_ABS/$INITRD" --kver "$KERNEL_VERSION" --kmoddir /lib/modules/"${KERNEL_VERSION}"
+
+        kernel="${KERNEL_IMAGE##*/}-${KERNEL_VERSION}.lz4"
+        compression="lz4"
```
is this compression type required by Coreboot ?

dorindabassey replied:
this is the compression method already being used by coreboot for the Initial kernel in the chromebook

martinezjavier replied:
I see. Maybe worth adding a comment then?

martinezjavier commented on 96-chromebook.install:
```diff
@@ -0,0 +1,62 @@
+            -b /boot/dtb/qcom/sc7180-trogdor-lazor-r3-kb.dtb \
+            -b /boot/dtb/rockchip/rk3399-gru-kevin.dtb\
+            -b /boot/dtb/rockchip/rk3399-gru-scarlet-inx.dtb \
+            "
```
Instead of hardcoding this, maybe we can pick all the files in `/boot/dtb/*` ?

dorindabassey replied:
yeah, this make sense but i don't think the fedora kernel generates all the device tree in /boot/dtb/* @eballetbo what d'you think.

martinezjavier replied:
but you are already relying on these DTB existing there... I just think that having those hardcoded here won't scale.

dorindabassey replied:
I think temporarily we can use these know dtb that the fedora kernel can generate then we can open a new ticket to figure out why the other existing dtb doesn't work with the FIT image. what d'you think?

martinezjavier commented on 96-chromebook.install:
```diff
@@ -0,0 +1,62 @@
+                kernel.itb
+
+	bootloader="boot_params"
+	echo "root=PARTUUID=%U/PARTNROFF=1 rootwait rw" > boot_params
```
No, instead just use the cmdline in `/proc/cmdline`. Or better:

```
if [[ -f /etc/kernel/cmdline ]]; then
    read -r -d '' -a BOOT_OPTIONS < /etc/kernel/cmdline
elif [[ -f /usr/lib/kernel/cmdline ]]; then
    read -r -d '' -a BOOT_OPTIONS < /usr/lib/kernel/cmdline
else
    declare -a BOOT_OPTIONS

    read -r -d '' -a line < /proc/cmdline
    for i in "${line[@]}"; do
        [[ "${i#initrd=*}" != "$i" ]] && continue
        [[ "${i#BOOT_IMAGE=*}" != "$i" ]] && continue
        BOOT_OPTIONS+=("$i")
    done
fi
```

like is done by `/usr/lib/kernel/install.d/90-loaderentry.install`.

dorindabassey replied:
noted

martinezjavier commented on 96-chromebook.install:
```diff
@@ -0,0 +1,62 @@
+		               --bootloader $bootloader \
+		               --vmlinuz kernel.itb \
+		               --arch arm
+	dd if=/boot/kernel.vboot of=/dev/sda1 bs=4M status=progress
```
we shouldn't hardcode `/dev/sda1` but instead figure out what is the kernel partition using `cgpt` or something like that.

dorindabassey replied:
Alright, thanks. will fix that

eballetbo commented on chromebook-setup.sh:
```diff
@@ -737,8 +741,8 @@ cmd_setup_fedora_kernel()
     mkdir ./tmpdir && cd ./tmpdir
-    rpm2cpio ../kernel-core-$kernel_version.rpm | cpio -idmv
-    rpm2cpio ../kernel-modules-$kernel_version.rpm | cpio -idmv
+    rpm2cpio ../kernel-core-$kernel_version.rpm | cpio -idm
```
This is an unrelated change. Please remove

eballetbo commented on chromebook-setup.sh:
```diff
@@ -737,8 +741,8 @@ cmd_setup_fedora_kernel()
-    rpm2cpio ../kernel-core-$kernel_version.rpm | cpio -idmv
-    rpm2cpio ../kernel-modules-$kernel_version.rpm | cpio -idmv
+    rpm2cpio ../kernel-core-$kernel_version.rpm | cpio -idm
+    rpm2cpio ../kernel-modules-$kernel_version.rpm | cpio -idm
```
Same here. Please remove that change

eballetbo commented on chromebook-setup.sh:
```diff
@@ -845,7 +849,7 @@ ensure_command realpath realpath
 ensure_command lz4 lz4
 ensure_command mkfs.ext4 e2fsprogs
-ensure_command mkimage u-boot-tools
+ensure_command mkimage uboot-tools
```
Hmm, so debian names this package as u-boot-tools whether fedora names it as uboot-tools. Not sure how e need to handle these cases but in any case I think it should be a separate patch. So I'd prefer do not change it in this PR.

dorindabassey replied:
okay, i see. fedora names it otherwise. I'd create a separate patch to fix that then

eballetbo reviewed (changes requested):
Thanks for the PR. I requested some minor comments, let me know if you're fine.

eballetbo commented on 96-chromebook.install:
```diff
@@ -0,0 +1,84 @@
+#!/usr/bin/bash
```
Can you prepend de commit message title with `chromebook-setup:`

`chromebook-setup: Extend Fedora to support kernel updates on Chromebooks`

Just to follow the same pattern used in all other commits.

dorindabassey replied:
ok will do thanks

eballetbo commented on chromebook-setup.sh:
```diff
@@ -742,6 +742,10 @@ cmd_setup_fedora_rootfs()
 
+    # echo "Inserting Kernel Update script"
+    sudo chmod 777 96-chromebook.install
+    sudo cp 96-chromebook.install ./tmpdir/root/usr/lib/kernel/install.d/
```
Instead of putting this file inside the source path, can we create a new directory called scripts and put the file there?

dorindabassey replied:
cool, make sense

eballetbo reviewed (changes requested):
Hi Dorinda, two minor comments, more cosmetic than problematic, other than that the patch looks good to me.

Commit:
chromebook-setup: Extend Fedora to support kernel updates on Chromebooks

Added a kernel-install script that creates an intramfs and FIT Image (including the ramdisk)
then the script flashes the kernel to the right partition in order to support kernel updates on the chromebook

Signed-off-by: Dorinda Bassey <dbassey@redhat.com>

eballetbo commented:
Merged. Many thanks for contributing.
